lib/agent-redirect.js

```diff
@@ -0,0 +1,199 @@
+    this.opts = opts
+    this.handler = handler
+    this.destroyed = false
+  }
+
+  destroy (err) {
+    this.destroyed = true
+
+    if (this.abort) {
+      this.abort()
+    }
+
+    this.handler.onError(err)
+  }
+
+  get (origin) {
+    return this[kAgent].get(origin)
+  }
+
+  onConnect (abort) {
+    if (this.destroyed) {
+      abort()
+    } else {
+      this.abort = abort
+      this.handler.onConnect(abort)
+    }
+  }
+
+  onUpgrade (statusCode, headers, socket) {
+    this.handler.onUpgrade(statusCode, headers, socket)
+  }
+
+  onError (error) {
+    this.handler.onError(error)
+  }
+
+  onHeaders (statusCode, headers, resume) {
+    if ([300, 301, 302, 303, 307, 308].indexOf(statusCode) === -1) {
+      return this.handler.onHeaders(statusCode, headers, resume)
+    }
+
+    this.location = parseLocation(headers)
+
+    if (!this.location) {
+      return this.handler.onHeaders(statusCode, headers, resume)
+    }
+
+    const { origin, pathname, search } = util.parseURL(new URL(this.location, this.opts.origin))
+    const path = search ? `${pathname || '/'}${search || ''}` : pathname
+
+    this.opts = { ...this.opts }
+
+    // Remove headers referring to the original URL.
+    // By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+    // https://tools.ietf.org/html/rfc7231#section-6.4
+    this.opts.headers = cleanRequestHeaders(this.opts.headers, statusCode === 303)
+    this.opts.path = path
+    this.opts.origin = origin
+
+    // https://tools.ietf.org/html/rfc7231#section-6.4.4
+    // In case of HTTP 303, always replace method to be either HEAD or GET
+    if (statusCode === 303 && this.opts.method !== 'HEAD') {
+      this.opts.method = 'GET'
+      this.opts.body = null
+    }
+  }
+
+  onData (chunk) {
+    if (!this.location) {
+      return this.handler.onData(chunk)
+    }
+  }
+
+  onComplete (trailers) {
+    if (this.location) {
+      this.agent.dispatch(this.opts, this.handler)
+    } else {
+      this.handler.onComplete(trailers)
+    }
+  }
+}
+
+class RedirectAgent {
+  constructor ({ maxRedirections = 10, agent } = {}) {
+    if (!Number.isInteger(maxRedirections) || maxRedirections < 0) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    if (agent != null && typeof agent.dispatch !== 'function') {
+      throw new InvalidArgumentError('invalid opts.agent')
+    }
+
+    this[kMaxRedirections] = maxRedirections
+    this[kAgent] = agent
+    this[kRedirectTag] = Symbol('tag')
+  }
+
+  dispatch (opts, handler) {
+    const { maxRedirections = this[kMaxRedirections] } = opts
+
+    if (maxRedirections != null && (maxRedirections < 0 || !Number.isInteger(maxRedirections))) {
+      throw new InvalidArgumentError('maxRedirections must be a positive number')
+    }
+
+    opts = { ...opts }
+
+    assert(opts[kRedirectTag] !== this[kRedirectTag], 'circular dispatch')
+    opts[kRedirectTag] = this[kRedirectTag]
```

Comment:
@mcollina wdyt?

Reply:
Can you clarify what is this for?

Reply:
It is possible to create a loop if you pass a redirect agent into a redirect agent.

Reply:
Oh, I see. Cool.